tests/test_native_mcm.py

```diff
@@ -0,0 +1,423 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for default qubit preprocessing."""
+from functools import reduce
+from typing import Sequence
+
+import numpy as np
+import pennylane as qml
+import pytest
+from conftest import LightningDevice, device_name
+from flaky import flaky
+from pennylane._device import DeviceError
+
+if not LightningDevice._new_API:
+    pytest.skip("Exclusive tests for new device API. Skipping.", allow_module_level=True)
+
+if not LightningDevice._CPP_BINARY_AVAILABLE:  # pylint: disable=protected-access
+    pytest.skip("No binary module found. Skipping.", allow_module_level=True)
+
+
+def validate_counts(shots, results1, results2):
+    """Compares two counts.
+
+    If the results are ``Sequence``s, loop over entries.
+
+    Fails if a key of ``results1`` is not found in ``results2``.
+    Passes if counts are too low, chosen as ``100``.
+    Otherwise, fails if counts differ by more than ``20`` plus 20 percent.
+    """
+    if isinstance(results1, Sequence):
+        assert isinstance(results2, Sequence)
+        assert len(results1) == len(results2)
+        for r1, r2 in zip(results1, results2):
+            validate_counts(shots, r1, r2)
+        return
+    for key1, val1 in results1.items():
+        val2 = results2[key1]
+        if abs(val1 + val2) > 100:
+            assert np.allclose(val1, val2, rtol=20, atol=0.2)
+
+
+def validate_samples(shots, results1, results2):
+    """Compares two samples.
+
+    If the results are ``Sequence``s, loop over entries.
+
+    Fails if the results do not have the same shape, within ``20`` entries plus 20 percent.
+    This is to handle cases when post-selection yields variable shapes.
+    Otherwise, fails if the sums of samples differ by more than ``20`` plus 20 percent.
+    """
+    if isinstance(shots, Sequence):
+        assert isinstance(results1, Sequence)
+        assert isinstance(results2, Sequence)
+        assert len(results1) == len(results2)
+        for s, r1, r2 in zip(shots, results1, results2):
+            validate_samples(s, r1, r2)
+    else:
+        sh1, sh2 = results1.shape[0], results2.shape[0]
+        assert np.allclose(sh1, sh2, rtol=20, atol=0.2)
+        assert results1.ndim == results2.ndim
+        if results2.ndim > 1:
+            assert results1.shape[1] == results2.shape[1]
+        np.allclose(np.sum(results1), np.sum(results2), rtol=20, atol=0.2)
+
+
+def validate_expval(shots, results1, results2):
+    """Compares two expval, probs or var.
```

Comment:
```suggestion
    """Compares two expval.
```

Reply:
The docstring is OK, I'll change the function name.